// @elastic/eui/lib/components/comment_list/comment_event.styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.euiCommentEventStyles = exports.euiCommentEventHeaderStyles = exports.euiCommentEventBorderColors = exports.euiCommentEventBodyStyles = void 0;
var _react = require("@emotion/react");
var _services = require("../../services");
var _global_styling = require("../../global_styling");
function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; } /*
                                                                                                                                                                                                                                                                                           * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
                                                                                                                                                                                                                                                                                           * or more contributor license agreements. Licensed under the Elastic License
                                                                                                                                                                                                                                                                                           * 2.0 and the Server Side Public License, v 1; you may not use this file except
                                                                                                                                                                                                                                                                                           * in compliance with, at your election, the Elastic License 2.0 or the Server
                                                                                                                                                                                                                                                                                           * Side Public License, v 1.
                                                                                                                                                                                                                                                                                           */
var euiCommentEventBorderColors = function euiCommentEventBorderColors(_ref3) {
  var euiTheme = _ref3.euiTheme,
    colorMode = _ref3.colorMode;
  var ratio = 0.6;
  return {
    warning: /*#__PURE__*/(0, _react.css)("border-color:", (0, _services.tintOrShade)(euiTheme.colors.warning, 0.4, colorMode), ";;label:warning;"),
    accent: /*#__PURE__*/(0, _react.css)("border-color:", (0, _services.tintOrShade)(euiTheme.colors.accent, ratio, colorMode), ";;label:accent;"),
    primary: /*#__PURE__*/(0, _react.css)("border-color:", (0, _services.tintOrShade)(euiTheme.colors.primary, ratio, colorMode), ";;label:primary;"),
    success: /*#__PURE__*/(0, _react.css)("border-color:", (0, _services.tintOrShade)(euiTheme.colors.success, ratio, colorMode), ";;label:success;"),
    danger: /*#__PURE__*/(0, _react.css)("border-color:", (0, _services.tintOrShade)(euiTheme.colors.danger, ratio, colorMode), ";;label:danger;"),
    subdued: /*#__PURE__*/(0, _react.css)("border-color:", euiTheme.border.color, ";;label:subdued;"),
    transparent: /*#__PURE__*/(0, _react.css)("border-color:", euiTheme.border.color, ";;label:transparent;"),
    plain: /*#__PURE__*/(0, _react.css)("border-color:", euiTheme.border.color, ";;label:plain;")
  };
};
exports.euiCommentEventBorderColors = euiCommentEventBorderColors;
var _ref2 = process.env.NODE_ENV === "production" ? {
  name: "yoyozp-euiCommentEvent",
  styles: "overflow:hidden;label:euiCommentEvent;"
} : {
  name: "yoyozp-euiCommentEvent",
  styles: "overflow:hidden;label:euiCommentEvent;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
var euiCommentEventStyles = function euiCommentEventStyles(euiThemeContext) {
  var euiTheme = euiThemeContext.euiTheme;
  return {
    euiCommentEvent: _ref2,
    border: /*#__PURE__*/(0, _react.css)("border-width:", euiTheme.border.width.thin, ";border-style:solid;border-radius:", euiTheme.border.radius.medium, ";;label:border;")
  };
};
exports.euiCommentEventStyles = euiCommentEventStyles;
var _ref = process.env.NODE_ENV === "production" ? {
  name: "ui5tce-euiCommentEvent__headerEvent",
  styles: "align-items:center;display:inline-flex;white-space:pre-wrap;flex-wrap:wrap;label:euiCommentEvent__headerEvent;"
} : {
  name: "ui5tce-euiCommentEvent__headerEvent",
  styles: "align-items:center;display:inline-flex;white-space:pre-wrap;flex-wrap:wrap;label:euiCommentEvent__headerEvent;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
var euiCommentEventHeaderStyles = function euiCommentEventHeaderStyles(euiThemeContext) {
  var euiTheme = euiThemeContext.euiTheme;
  return {
    euiCommentEvent__header: /*#__PURE__*/(0, _react.css)(";label:euiCommentEvent__header;"),
    border: /*#__PURE__*/(0, _react.css)((0, _global_styling.logicalCSS)('border-bottom-style', 'solid'), " ", (0, _global_styling.logicalCSS)('border-bottom-width', euiTheme.border.width.thin), ";;label:border;"),
    // Children
    euiCommentEvent__headerMain: /*#__PURE__*/(0, _react.css)("display:flex;flex:1;gap:", euiTheme.size.s, ";;label:euiCommentEvent__headerMain;"),
    euiCommentEvent__headerData: /*#__PURE__*/(0, _react.css)("display:flex;flex:1;align-items:center;flex-wrap:wrap;gap:", euiTheme.size.xs, ";;label:euiCommentEvent__headerData;"),
    euiCommentEvent__headerEventIcon: /*#__PURE__*/(0, _react.css)((0, _global_styling.logicalCSS)('margin-right', euiTheme.size.xs), ";;label:euiCommentEvent__headerEventIcon;"),
    euiCommentEvent__headerUsername: /*#__PURE__*/(0, _react.css)("font-weight:", euiTheme.font.weight.semiBold, ";;label:euiCommentEvent__headerUsername;"),
    euiCommentEvent__headerEvent: _ref,
    euiCommentEvent__headerActions: /*#__PURE__*/(0, _react.css)("display:flex;flex-wrap:wrap;gap:", euiTheme.size.xs, ";;label:euiCommentEvent__headerActions;")
  };
};
exports.euiCommentEventHeaderStyles = euiCommentEventHeaderStyles;
var euiCommentEventBodyStyles = function euiCommentEventBodyStyles(_ref4) {
  var euiTheme = _ref4.euiTheme;
  return {
    euiCommentEvent__body: /*#__PURE__*/(0, _react.css)(";label:euiCommentEvent__body;"),
    // types
    regular: /*#__PURE__*/(0, _react.css)("padding:", euiTheme.size.s, ";;label:regular;"),
    update: /*#__PURE__*/(0, _react.css)((0, _global_styling.logicalCSS)('padding-top', euiTheme.size.xs), ";;label:update;"),
    custom: /*#__PURE__*/(0, _react.css)(";label:custom;")
  };
};
exports.euiCommentEventBodyStyles = euiCommentEventBodyStyles;
