// @elastic/eui/optimize/lib/components/description_list/description_list_types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TYPES = exports.TEXT_STYLES = exports.ROW_GUTTER_SIZES = exports.COLUMN_GUTTER_SIZES = exports.CHILD_TYPES = exports.ALIGNMENTS = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

var CHILD_TYPES = ['row', 'inline', 'column'];
exports.CHILD_TYPES = CHILD_TYPES;
var TYPES = [].concat(CHILD_TYPES, ['responsiveColumn']);
exports.TYPES = TYPES;
var ALIGNMENTS = ['center', 'left'];
exports.ALIGNMENTS = ALIGNMENTS;
var TEXT_STYLES = ['normal', 'reverse'];
exports.TEXT_STYLES = TEXT_STYLES;
var ROW_GUTTER_SIZES = ['s', 'm'];
exports.ROW_GUTTER_SIZES = ROW_GUTTER_SIZES;
var COLUMN_GUTTER_SIZES = ['s', 'm'];
exports.COLUMN_GUTTER_SIZES = COLUMN_GUTTER_SIZES;
